File: apps/web/lib/api/perceptron.ts
import { apiPost } from "@/lib/api/client";
import type {
  DatasetName,
  PerceptronResponse,
  PerceptronState,
} from "@/lib/types/perceptron";

export async function initializePerceptron(
  dataset: DatasetName,
  learningRate = 0.1,
): Promise<PerceptronResponse> {
  return apiPost<PerceptronResponse>("/api/perceptron/initialize", {
    dataset,
    learning_rate: learningRate,
  });
}

export async function trainPerceptron(
  state: PerceptronState,
  steps: number,
): Promise<PerceptronResponse> {
  return apiPost<PerceptronResponse>("/api/perceptron/train", {
    state,
    steps,
  });
}

File: apps/api/app/routers/perceptron.py
from fastapi import APIRouter

from app.schemas.perceptron import InitializeRequest, TrainRequest
from app.services.perceptron_service import initialize_state, train_state

router = APIRouter(prefix="/api/perceptron", tags=["perceptron"])


@router.post("/initialize")
def initialize(req: InitializeRequest) -> dict[str, object]:
    return initialize_state(req.dataset, req.learning_rate)


@router.post("/train")
def train(req: TrainRequest) -> dict[str, object]:
    state = req.state
    return train_state(
        dataset=state.dataset,
        weights=state.weights,
        bias=state.bias,
        learning_rate=state.learning_rate,
        epoch=state.epoch,
        steps=req.steps,
    )
